File: src/main/java/kr/pe/ecmaxp/thumbsj/Registers.java
package kr.pe.ecmaxp.thumbsj;

import kr.pe.ecmaxp.thumbsj.helper.BitConsts;
import kr.pe.ecmaxp.thumbsj.helper.RegisterIndex;

public class Registers
{
    private int[] _regs;

    @SuppressWarnings("WeakerAccess")
    public Registers()
    {
        _regs = new int[RegisterIndex.CPSR + 1];
        _regs[RegisterIndex.CPSR] = (
                BitConsts.FT | // set Thumb
                        BitConsts.FZ // set Zero
        );
    }

    @SuppressWarnings("WeakerAccess")
    public int[] load()
    {
        return _regs.clone();
    }

    @SuppressWarnings("WeakerAccess")
    public void store(int[] regs)
    {
        _regs = regs.clone();
    }

    public Registers copy()
    {
        Registers registers = new Registers();
        int[] regs = load();
        registers.store(regs);
        return registers;
    }

    public void set(int index, int value)
    {
        _regs[index] = value;
    }

    public int get(int index)
    {
        return _regs[index];
    }
}

File: src/main/java/kr/pe/ecmaxp/thumbsj/Registers.kt
package kr.pe.ecmaxp.thumbsj

import kr.pe.ecmaxp.thumbsj.helper.BitConsts
import kr.pe.ecmaxp.thumbsj.helper.RegisterIndex

class Registers {
    private var _regs: IntArray? = IntArray(RegisterIndex.CPSR + 1)

    init {
        _regs!![RegisterIndex.CPSR] = BitConsts.FT or // set Thumb
                BitConsts.FZ // set Zero
    }

    fun load(): IntArray {
        return _regs!!.clone()
    }

    fun store(regs: IntArray) {
        _regs = regs.clone()
    }

    fun copy(): Registers {
        val registers = Registers()
        val regs = load()
        registers.store(regs)
        return registers
    }

    operator fun set(index: Int, value: Int) {
        _regs!![index] = value
    }

    operator fun get(index: Int): Int {
        return _regs!![index]
    }
}
